Added orientation control to OSC

studywolf opened the pull request:
Added the ability to specify target Euler angles for the end-effector to the operational space controller. It now accepts an additional parameter, `ctrlr_dof` which is a 6D list specifying which dimensions of `[x, y, z, alpha, beta, gamma]` you would like to control. NOTE: if the arm under control has fewer DOF than the number of DOF specified for control you will get poor behaviour (usually oscillation between several DOF targets).

Also changed `target_pos` -> `target` in all the controllers, and the input `target` and `target_vel` for the OSC must now be 6D vectors.

Cleaned up the examples, reducing code (removing generation calls outside of main control loop) and removing unused tracking code, updated comments, added two more examples for both the PyGame and VREP folders (`orientation_control_osc` and `position_and_orientation_control_osc`), renamed `reaching_*` -> `position_control_*`.

Also I removed the commented code from the OSC `null space controller` and removed `use_dJ` as an option.

p3jawors commented on examples/PyGame/dynamics_adaptation_sliding.py:
```diff
@@ -31,6 +32,7 @@
     n_output=robot_config.N_JOINTS,
     pes_learning_rate=1e-2, backend='nengo')
```
might want to try a lower learning rate, depending on where the target is the arm can become unstable

studywolf replied:
noted! will reduce

p3jawors commented on examples/VREP/dynamics_adaptation_osc.py:
```diff
@@ -68,10 +79,6 @@
 
         # add an additional force for the controller to adapt to
         extra_gravity = robot_config.g(feedback['q']) * 2
```
I don't think you need the `*2` since just adding gravity again is already doing double gravity, however it doesn't really matter, just takes longer to adapt

studywolf replied:
yeah it's more of a visible effect so i added it in

p3jawors commented on examples/VREP/orientation_control_osc.py:
```diff
@@ -0,0 +1,70 @@
+            interface.get_orientation('target')])
+
+        u = ctrlr.generate(
+            q=feedback['q'],
```
is this just setting the target pos and orientation to the what the target is? When I run the sim I don't see it reaching to anything, and the ball labelled 'target' is below the floor surface.

studywolf replied:
oh right, need to change the default position of target haha, and yes that's what it's doing

p3jawors commented on examples/VREP/position_and_orientation_control_osc.py:
```diff
@@ -0,0 +1,96 @@
+target_angles_track = []
+
+# control (x, y, beta, gamma) out of [x, y, z, alpha, beta, gamma]
+ctrlr_dof = [True, True, False, False, True, True]
```
How can you reach in 3d space if you don't control z? I tried moving the target above the floor manually and it looks like it's reaching towards it, but can't get there because it can't change z. It ends up about 40cm above the target

studywolf replied:
oop yes i'll change the target location in the VREP file. but what's happening here is that we're specifying that we don't care about z, so it'll never match that. what matters here are the x, y, and beta, gamma angles. while the angles matching can't be seen visually they show up in the plot afterwards

studywolf commented on abr_control/controllers/osc.py:
```diff
@@ -12,7 +14,9 @@ class OSC(controller.Controller):
     kp : float, optional (Default: 1)
-        proportional gain term
+        proportional gain term on task space position
+    kp : float, optional (Default: 1)
```
ah dang this is old, there's only `kp`, i'll remove the second comment (which should have been `ko` yeah)

studywolf commented on abr_control/controllers/osc.py:
```diff
@@ -30,46 +34,41 @@ class OSC(controller.Controller):
     use_C : boolean, optional (Default: False)
         calculate and compensate for the Coriolis and
         centripetal effects of the arm
-    use_dJ : boolean, optional (Default: False)
```
yeah i'm not actually sure that it was ever appropriate, we'll still leave it in the base config but need to do more reading before to find out how / if we should incorporate it in the osc

studywolf commented on abr_control/controllers/osc.py:
```diff
@@ -78,30 +77,46 @@ def generate(self, q, dq,
-            desired joint angles [radians]
-        target_vel : float numpy.array, optional (Default: numpy.zeros)
-            desired joint velocities [radians/sec]
+        target: 6 dimensional float numpy.array
```
yeah, i was torn about this. basically the issue is that if we do that then the code below gets more complicated. I figured it was an easy enough thing for the user to do outside the call, with the hopes of keeping the code in this function at a minimum for speed

studywolf commented on abr_control/controllers/osc.py:
```diff
@@ -78,30 +77,46 @@ def generate(self, q, dq,
+        ctrlr_dof : 6D list of boolean, optional (Default: position control)
+            specifies which task space degrees of freedom are to be controlled
+            [x, y, z, alpha, beta, gamma] (Default: [x, y, z]
+            NOTE: if more ctrlr_dof are specified than degrees of freedom in
```
i was thinking about this and wasn't sure, because i think it'll depend also on the configuration of the robot. iirc it's related to the form of the Jacobian, another thing that will require more reading before implementation. i feeeeeel like having a note here is enough atm that we don't have to restrict the user ... not sure

studywolf commented on examples/PyGame/avoid_joint_limits.py:
```diff
@@ -1,6 +1,6 @@
-Running the three joint arm with the PyGame display, using an exponential
-additive signal when to push away from joints and  avoid self collision.
+Running operational space control with the PyGame display, using an exponential
+additive signal when to push away from joints and avoid self collision.
```
good point, i'll update the comment

studywolf commented:
OK, made the changes, can you please confirm the targets are now reachable in the VREP UR5 and Jaco2 files?

p3jawors commented:
LGTM, vrep targets are reachable and the pygame examples that were updated still run

p3jawors commented on abr_control/controllers/osc.py:
```diff
@@ -78,30 +79,46 @@ def generate(self, q, dq,
+        target: 6 dimensional float numpy.array
+            desired task space position and orientation [meters, radians]
+        target_vel : float 6D numpy.array, optional (Default: None)
+            desired task space velocities [meters/sec, radians/sec]
```
shouldn't this be just rad/sec? m/s would be if we were setting the desired velocity in task space.

studywolf replied:
this is in task space

p3jawors commented on examples/VREP/position_control_osc.py:
```diff
@@ -28,39 +29,52 @@
+        rc_angles = transformations.euler_from_matrix(rc_matrix, axes='sxyz')
+
+        interface.set_xyz('object', robot_config.Tx(name, q=feedback['q']))
+        interface.set_orientation('object', rc_angles)
```
Why are we setting an orientation if we are only controlling x, y, and z. Alpha, beta, and gamma are all `False` in `ctrlr_dof`

p3jawors commented:
ISSUE running dmps_path_planning
Traceback (most recent call last):
  File "dmps_path_planning.py", line 26, in <module>
    dmps = pydmps.DMPs_rhythmic(n_dmps=2, n_bfs=50, dt=.01)
NameError: name 'pydmps' is not defined

- There is a try except that should stop this and link to your github
not sure why it isn't being triggered

p3jawors commented:
ISSUE with all examples using oisc.py
pygame
avoid-joint-limits
avoid-obstacles
dynamics-adaptation-osc
orientation-control-osc
path-planning-linear
path-planning-second-order
position-and-orientation-control-osc
position-control-osc
dmp-path-planning
Traceback (most recent call last):
  File "avoid_joint_limits.py", line 20, in <module>
    ctrlr = OSC(robot_config, kp=100, vmax=10)
  File "/home/pawel/src/abr_control/abr_control/controllers/osc.py", line 57, in __init__
    self.kv = np.sqrt(kp+ko) if kv is None else kv
TypeError: unsupported operand type(s) for +: 'int' and 'NoneType'

- The previous lines should prevent this situation from happening, but
a printout confirms that ko=None

VREP
dynamics-adaptation-osc
obstacle-avoidance
orientation-control-osc
position-and-orienation-control-osc
position-control-osc

Traceback (most recent call last):
  File "avoid_joint_limits.py", line 20, in <module>
    ctrlr = OSC(robot_config, kp=100, vmax=10)
  File "/home/pawel/src/abr_control/abr_control/controllers/osc.py", line 57, in __init__
    self.kv = np.sqrt(kp+ko) if kv is None else kv
TypeError: unsupported operand type(s) for +: 'int' and 'NoneType'

floating
Traceback (most recent call last):
  File "floating.py", line 33, in <module>
    feedback = interface.get_feedback()
  File "/home/pawel/src/abr_control/abr_control/interfaces/vrep.py", line 272, in get_feedback
    'return code ', _)
Exception: ('Error retrieving joint angle, return c

studywolf commented:
@p3jawors did a rebase and cleanup of the history, could you look this over again?

studywolf commented:
OK, all tests passing and examples running. @p3jawors can you look this over again? I rebased again to the current master and fixed the history botch I did Thursday.

p3jawors commented on abr_control/controllers/osc.py:
```diff
@@ -21,6 +24,11 @@ class OSC(Controller):
+        specifies which task space degrees of freedom are to be controlled
+        [x, y, z, alpha, beta, gamma] (Default: [x, y, z]
+        NOTE: if more ctrlr_dof are specified than degrees of freedom in
+        the robotic system, the controller will perform poorly
```
should we add an `assert sum(ctrlr_dof) <= robot_config.N_JOINTS`? If not an assertion, maybe print a warning?

studywolf replied:
hmm i wouldn't want to make it an assertion, but a warning is a good idea. i'll add this now

p3jawors commented on examples/VREP/position_control_osc.py:
```diff
@@ -35,34 +39,41 @@
+        name = 'joint1'
+        vrep_angles = interface.get_orientation('UR5_%s' % name)
+        rc_matrix = robot_config.R(name, feedback['q'])
+        rc_angles = transformations.euler_from_matrix(rc_matrix, axes='sxyz')
```
Do we need lines 59 and 60?

studywolf replied:
good catch, those can be taken out

p3jawors commented:
- Added some docstrings
- fixed 3d plotting in an example
- Commit squashed
- examples all run
- style checks pass

LGTM
